src/main/java/com/google/googleidentity/oauth2/util/OAuth2Utils.java

```diff
@@ -31,6 +36,32 @@
     private static final String USER_SESSION = "user_session";
     private static final String CLIENT_SESSION = "client_session";
 
+    /**
+     * Return oauth2 exception error through httpResponse.
+     */
+    public static void returnHttpError(
+            HttpServletResponse response, OAuth2Exception exception) throws IOException {
+        response.setStatus(exception.getCode());
+        response.setContentType("application/json");
+
+        JSONObject json =  new JSONObject();
+        json.appendField("error", exception.getErrorType().get());
+        if(exception.getErrorInfo().isPresent()){
```

Comment:
In general, I am not sure which is better,
1.  make the exception class more "internal" and use a helper class to map the exception into http response
2. as you are implementing now, make exception to contain http response details directly.
I am lean to #1 since it makes the later implementation more clear and less redundant code, because I think there would be many places returning the same http error, so you don't have to repeatedly filling all information into exception class every time.
@jstoops-at-google WDYT?

Reply:
I agree that it'd be nice to decouple json map stuff from throwing exceptions. More specific exception subclasess like InvalidRequestException() will help. If you had some common interface or superclass, then the exceptions could all return error, error_description, and additional stuff in the same way that OAuth2Exception does here.